### PR TITLE
Unify Rydberg and Transmon Hamiltonians to ``HardwareHamiltonian``

### PR DESCRIPTION
**Pull request by Qottmann**
Based on https://github.com/PennyLaneAI/pennylane/pull/3896, https://github.com/PennyLaneAI/pennylane/pull/3749 and https://github.com/PennyLaneAI/pennylane/pull/3866, build Hamiltonians for transmon systems

Unifying hardware hamiltonians for both transmon and rydberg systems
Most changes are simple renaming and moving, there are no code changes compared to https://github.com/PennyLaneAI/pennylane/pull/3866. The biggest change is in the doc string.

Related forward work: https://github.com/PennyLaneAI/pennylane/pull/3936 to add `transmon_interaction`

### CODE REVIEW

**Comment by Qottmann** [comment]
[sc-35359]

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -13,10 +13,11 @@
     Rydberg atoms.
   * A new user-facing `rydberg_interaction` function is added, which returns a `RydbergHamiltonian` containing
     the Hamiltonian of the interaction of all the Rydberg atoms.
-  * A new user-facing `rydberg_drive` function is added, which returns a `RydbergHamiltonian` containing
-    the Hamiltonian of the interaction between a driving laser field and a group of atoms.
+  * A new user-facing `drive` function is added, which returns a `HardwareHamiltonian` containing
+    the Hamiltonian of the interaction between a driving laser field and a group of qubits.
```
Comment (lillian542) [inline doc/releases/changelog-dev.md:17]:
```suggestion
    the Hamiltonian of the interaction between a driving electro-magnetic field and a group of qubits.
```

File: pennylane/pulse/hardware_hamiltonian.py [hunk 2]
```diff
@@ -157,75 +68,145 @@ def rydberg_drive(amplitude, phase, detuning, wires):
...
+
+        with amplitude :math:`\Omega`, phase :math:`\phi` and drive frequency :math:`\nu` of the electromagnetic field, as well as the qubit frequency :math:`\omega_q`.
+        We can move to the rotating frame of the driving field by applying :math:`U = e^{-i\nu t \sigma^z / 2}` which yields the new Hamiltonian
+
+        .. math::
+            H = \frac{1}{2} \Omega(t) \sum_{j \in \text{wires}} \left(e^{i \phi(t)} \sigma^+_j + e^{-i \phi(t)} \sigma^-_j \right)
+            - \frac{1}{2} (\nu - \omega_q) \sum_{j \in \text{wires}} \sigma^z_j
+
+        We can define :math:`\Delta = \nu - \omega_q` to arrive at the definition above. Note that a potential anharmonicity term, as is common for transmon systems,
+        is unaffected by this transformation. (TODO link transmon anharmonicity constructor)
```
Comment (lillian542) [inline pennylane/pulse/hardware_hamiltonian.py:165]: As long as we're limited to d=2, I would think the anharmonicity term is meaningless. Do we want to include discussion of this term from the get-go, or is it something we should add to the documentation when we add qutrit/qudit support?

File: pennylane/pulse/hardware_hamiltonian.py [hunk 3]
```diff
@@ -157,75 +68,145 @@ def rydberg_drive(amplitude, phase, detuning, wires):
...
+
+    .. details::
+        **Neutral Atom Rydberg systems**
+
+        In neutral atom systems for quantum computation and quantum simulation, a Rydberg transition is driven by an optical laser (frequency in MHz).
+        The interaction between different atoms is given by the :func:`rydberg_interaction`, for which we pass the atomic coordinates (in µm),
+        here arrange in a square of length :math:`4\my m`.
```
Comment (lillian542) [inline pennylane/pulse/hardware_hamiltonian.py:172]:
```suggestion
        here arrange in a square of length :math:`4 \mu m`.
```

**Review by lillian542 — COMMENTED** [review]
New names seem good to me :)

File: pennylane/pulse/hardware_hamiltonian.py (hunk 2 above)
Comment (Qottmann) [reply pennylane/pulse/hardware_hamiltonian.py:165]: I think it makes sense to mention to not confuse users expecting it, but yeah the actual implementation is currently not necessary, good point!

**Review by lillian542 — APPROVED** [review]
I left a few comments but overall looks good to me! I'll get started on the corresponding changes to the AWS device :)

File: doc/releases/changelog-dev.md [hunk 4]
```diff
@@ -13,10 +13,11 @@
     Rydberg atoms.
   * A new user-facing `rydberg_interaction` function is added, which returns a `RydbergHamiltonian` containing
     the Hamiltonian of the interaction of all the Rydberg atoms.
-  * A new user-facing `rydberg_drive` function is added, which returns a `RydbergHamiltonian` containing
-    the Hamiltonian of the interaction between a driving laser field and a group of atoms.
+  * A new user-facing `drive` function is added, which returns a `HardwareHamiltonian` containing
```
[the remark this answers is not in the document]
Comment (Qottmann) [reply doc/releases/changelog-dev.md:16]: Good point, I think we should actually _only_ mention `ParametrizedHamiltonian`, though this also works 👍

File: pennylane/pulse/hardware_hamiltonian.py [hunk 5]
```diff
@@ -157,75 +68,146 @@ def rydberg_drive(amplitude, phase, detuning, wires):
...
+            - \frac{1}{2} (\nu - \omega_q) \sum_{j \in \text{wires}} \sigma^z_j
+
+        We can define :math:`\Delta = \nu - \omega_q` to arrive at the definition above. Note that a potential anharmonicity term,
+        as is common for transmon systems when taking into account higher energy levels,
+        is unaffected by this transformation.
+
+    .. details::
+        **Neutral Atom Rydberg systems**
+
+        In neutral atom systems for quantum computation and quantum simulation, a Rydberg transition is driven by an optical laser (frequency in MHz).
```
[the remark this answers is not in the document]
Comment (Qottmann) [reply pennylane/pulse/hardware_hamiltonian.py:171]: Oh thanks that actually gave me a little epiphany here!

[the remark this answers is not in the document]
Comment (Qottmann) [reply pennylane/pulse/hardware_hamiltonian.py:171]: Yes that is great!

File: pennylane/pulse/hardware_hamiltonian.py [hunk 6]
```diff
@@ -28,127 +26,40 @@
...
     Args:
-        amplitude (Union[float, Callable]): float or callable returning the amplitude (in MHz) of a
-            laser field
-        phase (Union[float, Callable]): float or callable returning the phase (in radians) of the laser field
-        detuning (Union[float, Callable]): float or callable returning the detuning (in MHz) of a
-            laser field
-        wires (Union[int, List[int]]): integer or list containing wire values for the Rydberg atoms that
-            the laser field acts on
+        amplitude (Union[float, Callable]): float or callable returning the amplitude of a
+            electromagnetic field
```
Comment (mudit2812) [inline pennylane/pulse/hardware_hamiltonian.py:53]:
```suggestion
        amplitude (Union[float, Callable]): float or callable returning the amplitude of an
            electromagnetic field
```

File: pennylane/pulse/hardware_hamiltonian.py [hunk 7]
```diff
@@ -28,127 +26,40 @@
...
+        phase (Union[float, Callable]): float or callable returning the phase (in radians) of the electromagnetic field
+        detuning (Union[float, Callable]): float or callable returning the detuning of a
+            electromagnetic field
```
Comment (mudit2812) [inline pennylane/pulse/hardware_hamiltonian.py:56]:
```suggestion
        detuning (Union[float, Callable]): float or callable returning the detuning of an
            electromagnetic field
```

**Review by mudit2812 — APPROVED** [review]
Thanks Korbinian! This looks great, and the documentation for `drive` is very thorough.

**Pull request merged**